KEP-3866: kube-proxy nftables mode (#3824)
kubernetes/enhancements: keps/sig-network/3866-nftables-proxy/README.md

go into those tables. Other system components (eg, firewalld) should
ignore our table, so we should not need to worry about watching for
other people deleting our rules like we have to in the `iptables`
backend.

(In theory, instead of creating one table each in the `ip` and `ip6`
families, we could create a single table in the `inet` family and put
both IPv4 and IPv6 chains/rules there. However, this wouldn't really
result in much simplification, because we would still need separate
sets/maps to match IPv4 addresses and IPv6 addresses. (There is no
data type that can store/match either an IPv4 address or an IPv6
address.) Furthermore, because of how Kubernetes Services evolved in
parallel with the existing kube-proxy implementation, we have ended up
with a dual-stack Service semantics that is most easily implemented by
handling IPv4 and IPv6 completely separately anyway.)

#### Communicating with the kernel nftables subsystem

At least initially, we will use the `nft` command-line tool to read
and write rules, much like how we use command-line tools in the
`iptables` and `ipvs` backends. However, the `nft` tool is mostly just

> a thin wrapper around `libnftables`, and it would be possible to use

Reviewer: we need to update the image for these new dependencies or create a new one
https://github.com/kubernetes/release/tree/master/images/build/distroless-iptables
```
 ldd /usr/sbin/nft
        linux-vdso.so.1 (0x00007ffc8cbcb000)
        libnftables.so.1 => /lib/x86_64-linux-gnu/libnftables.so.1 (0x00007feeb52d0000)
        libedit.so.2 => /lib/x86_64-linux-gnu/libedit.so.2 (0x00007feeb5296000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007feeb506e000)
        libmnl.so.0 => /lib/x86_64-linux-gnu/libmnl.so.0 (0x00007feeb5066000)
        libnftnl.so.11 => /lib/x86_64-linux-gnu/libnftnl.so.11 (0x00007feeb5034000)
        libxtables.so.12 => /lib/x86_64-linux-gnu/libxtables.so.12 (0x00007feeb5021000)
        libjansson.so.4 => /lib/x86_64-linux-gnu/libjansson.so.4 (0x00007feeb500f000)
        libgmp.so.10 => /lib/x86_64-linux-gnu/libgmp.so.10 (0x00007feeb4f8d000)
        libtinfo.so.6 => /lib/x86_64-linux-gnu/libtinfo.so.6 (0x00007feeb4f5b000)
        libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007feeb4f43000)
        /lib64/ld-linux-x86-64.so.2 (0x00007feeb53b5000)
        libmd.so.0 => /lib/x86_64-linux-gnu/libmd.so.0 (0x00007feeb4f36000)
```

Author: (note that `iptables-nft` already depends on half of those things too... it looks like nft should add about 1M to the image)

```
set hairpin {
  type ipv4_addr . ipv4_addr;
  elements {
    10.180.0.4 . 10.180.0.4,
    10.180.0.5 . 10.180.0.5,
    ...
  }
}

chain ... {
  ...
  ip saddr . ip daddr @hairpin jump mark_for_masquerade
}
```

More efficiently, if nftables eventually got the ability to call eBPF
programs as part of rule processing (like iptables's `-m ebpf`) then
we could write a trivial eBPF program to check "source IP equals
destination IP" and then call that rather than needing the giant set
of redundant IPs.

If we do this, then we don't need the per-endpoint hairpin check
rules. If we could also get rid of the per-endpoint affinity-updating
rules, then we could get rid of the per-endpoint chains entirely,
since `dnat to ...` is an allowed vmap verdict:

```
chain svc_4SW47YFZTEDKD3PK {
  # FIXME handle affinity somehow

  # Send to random endpoint
  random mod 2 vmap {
    0 : dnat to 10.180.0.4:8000
    1 : dnat to 10.180.0.5:8000
  }
}
```

With the current set of nftables functionality, it does not seem
possible to do this (in the case where affinity is in use), but future
features may make it possible.

It is not yet clear what the tradeoffs of such rewrites are, either in

> terms of runtime performance, or of admin/developer-comprehensibility

Reviewer: yeah, I was just going to ask this

ruleset to determine how they would behave.

The `cmd/kube-proxy/app` tests mostly only test configuration parsing,
and we will extend them to understand the new mode and its associated
configuration options, but there will not be many changes made there.

<!--
Additionally, for Alpha try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>
The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

- `<package>`: `<date>` - `<test coverage>`

##### Integration tests

Kube-proxy does not have integration tests.

##### e2e tests

Most of the e2e testing of kube-proxy is backend-agnostic. Initially,
we will need a separate e2e job to test the nftables mode (like we do
with ipvs). Eventually, if nftables becomes the default, then this
would be flipped around to having a legacy "iptables" job.

> The handful of e2e tests that specifically examine iptables rules will

Reviewer: I don't know if this still exist, but is not great for e2e test to have this implementation dependencies, I rather rewrite them to focus on the expected behavior or drop them entirely, I think you already did that for some of them

Author: Interesting... I had thought there were more of these. I guess they mostly got fixed. (You're right, [I did just remove one of them](https://github.com/kubernetes/kubernetes/pull/113335/commits/c6cc056675314ef7f74e4b318d39eadd73b2de40).) There is only one remaining test that parses iptables rules, [`"It should recreate its iptables rules if they are deleted [Disruptive]"`](https://github.com/kubernetes/kubernetes/blob/v1.26.1/test/e2e/network/networking.go#L551), which tests 2 things:
1. kubelet recreates `KUBELET-IPTABLES-HINT` if it is deleted (which is iptables-specific and should stick around regardless of kube-proxy, for as long as we expect third-party components to use iptables).
2. Services eventually start working again even if you delete all `KUBE-*` iptables chains. It doesn't actually test that they _stop_ working after you delete those chains, so it would still pass in nftables mode, it just wouldn't test anything interesting. (And as discussed in the KEP text, we probably don't need to do monitoring for nftables.)

There are also a few tests that manually invoke iptables to break the network via [`TestUnderTemporaryNetworkFailure`](https://github.com/kubernetes/kubernetes/blob/v1.27.0-alpha.2/test/e2e/framework/network/utils.go#L1078) (plus [one that doesn't use that function but should](https://github.com/kubernetes/kubernetes/blob/v1.27.0-alpha.2/test/e2e/apimachinery/etcd_failure.go#L63)). This will eventually need to be updated to have an nftables-based version, but that's independent of kube-proxy.

I think that's it? Yay us.

Author: filed https://github.com/kubernetes/kubernetes/issues/119047 about TestUnderTemporaryNetworkFailure

need to be updated to be able to work with either backend.

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

- <test>: <link to test coverage>

> 

Reviewer: We need a section for performance test, we need to check that at least these network metrics we are gathering in scalability improve, since is the main driven of this 

https://github.com/kubernetes/kubernetes/issues/115720

Reviewer 2: Agreed

Author: (I added the "Scalability tests" section immediately after this, in response to Antonio's original comment, though it's still just "TBD", because I don't really know a lot about our perf/scale testing.)

[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The default kube-proxy implementation on Linux is currently based on
iptables. IPTables was the preferred packet filtering and processing
system in the Linux kernel for many years (starting with the 2.4
kernel in 2001). However, problems with iptables led to the
development of a successor, nftables, first made available in the 3.13
kernel in 2014, and growing increasingly featureful and usable as a
replacement for iptables since then. Development on iptables has
mostly stopped, with new features and performance improvements
primarily going into nftables instead.

This KEP proposes the creation of a new official/supported nftables
backend for kube-proxy. While it is hoped that this backend will
eventually replace both the `iptables` and `ipvs` backends and become
the default kube-proxy mode on Linux, that replacement/deprecation
would be handled in a separate future KEP.

## Motivation

There are currently two officially supported kube-proxy backends for
Linux: `iptables` and `ipvs`. (The original `userspace` backend was

> deprecated several releases ago and removed from the tree in 1.25.)

Reviewer 3: ```suggestion
deprecated several releases ago and removed from the tree in 1.26.)
```
I think that's right; at least, the v1.25 docs still list the userspace mode.

The `iptables` mode of kube-proxy is currently the default, and it is
generally considered "good enough" for most use cases. Nonetheless,

> there are good arguments for replacing it with a new `nftables` mode.

Reviewer 3: Did you mean:
```suggestion
there are good arguments for adding a new `nftables` mode and eventually
removing the `iptables` mode.
```

Author: :man_shrugging: There are arguments for replacing it, and there are separate arguments for doing it slowly, which I talk about some later, but everyone already knows the arguments for "don't break the world" already... :slightly_smiling_face:

> ### The iptables kernel subsystem has unfixable performance problems

Reviewer 3: ```suggestion
### iptables kernel subsystem performance
```

Author: I think the fact that they are unfixiable is relevant to the discussion, in that this closes off some other alternatives

    NAT-based approach would work for IPv6 if [the client explicitly
    set a non-localhost source IP]. (Presumably it would also work in
    the IPv4 case without needing `route_localnet` in that case.) By
    the time the packet has reached the network stack though, it is
    probably too late to change the source IP? (OTOH, using eBPF to
    intercept the `connect()` call would probably work.)

  - It is possible that using the nftables `fwd` command from a rule
    on a table in the `netdev` family would let us move the packet
    around in a way that bypassed the kernel checks that otherwise
    make this not work?

  - We could resurrect enough of the userspace proxy to use it just
    for localhost NodePort connections... :vomiting_face:

      - (Alternatively: having a listening port on localhost, but then
        after accepting a connection, using eBPF to splice it directly
        to the destination server socket.)

It's not clear what the use cases for localhost NodePorts are. Tim's
[initial implementation of NodePorts in the iptables proxy] set
`route_localnet`, so localhost nodeports seem to have been considered
an expected feature already at that point.

>   - In kubernetes #90236, Tim commented that one use case is

Reviewer 3: ```suggestion
  - In kubernetes issue [#90236](https://github.com/kubernetes/kubernetes/issues/90236),
    Tim Hockin commented that one use case is
```

[connecting to a docker registry without needing to set up certificates]: https://github.com/kubernetes/kubernetes/issues/90236#issuecomment-622142806
[the only use-case I have heard]: https://github.com/kubernetes/kubernetes/pull/92938#issuecomment-742985382
[use the node IP instead]: https://github.com/contiv/vpp/pull/1434

#### NodePort Addresses

In addition to the localhost issue, iptables kube-proxy defaults to
accepting NodePort connections on all local IPs, which has effects
varying from intended-but-unexpected ("why can people connect to
NodePort services from the management network?") to clearly-just-wrong
("why can people connect to NodePort services on LoadBalancer IPs?")

The nftables proxy should default to only opening NodePorts on a
single interface, probably the interface with the default route by
default. (Ideally, you really want it to accept NodePorts on the
interface that holds the route to the cloud load balancers, but we
don't necessarily know what that is ahead of time.) Admins can use
`--nodeport-addresses` to override this.

#### Behavior of service IPs

> Tim has suggested that service IPs should reject connections on ports

Reviewer 3: ```suggestion
Tim Hockin has suggested that service IPs should reject connections on ports
```
We have multiple Tims and they don't support selecting based on annotations.

Author: There's sort of an implied `<<UNRESOLVED>>` around the entire "Changes from iptables" section; we'll want to resolve a lot of details here before merging. (And to the extent that we don't, yes, I'll fix things up before then.)

(both IPv4 and IPv6), `arp`, `bridge`, or `netdev`), but unlike in
iptables, you can have both "filter"-type chains and "NAT"-type chains
in the same table.

So, we will create a single `kube_proxy` table in the `ip` family, and
another in the `ip6` family. All of our chains, sets, maps, etc, will
go into those tables. Other system components (eg, firewalld) should
ignore our table, so we should not need to worry about watching for
other people deleting our rules like we have to in the `iptables`
backend.

(In theory, instead of creating one table each in the `ip` and `ip6`
families, we could create a single table in the `inet` family and put
both IPv4 and IPv6 chains/rules there. However, this wouldn't really
result in much simplification, because we would still need separate
sets/maps to match IPv4 addresses and IPv6 addresses. (There is no
data type that can store/match either an IPv4 address or an IPv6
address.) Furthermore, because of how Kubernetes Services evolved in
parallel with the existing kube-proxy implementation, we have ended up
with a dual-stack Service semantics that is most easily implemented by
handling IPv4 and IPv6 completely separately anyway.)

> 

Reviewer 3: `inet` might be useful if we want to make a `KUBE-NFTABLES-HINT` table.

Author: We won't need that; `KUBE-IPTABLES-HINT` is needed because iptables has two modes and containerized apps need some way to figure out which mode the host filesystem is configured to use. But there's no corresponding thing with nftables. (And, if there was, then the hint would have to be set by kubelet, not kube-proxy anyway.)

Reviewer 3: It's also a hint about whether `kube-proxy` is using iptables or nftables. At least, I think it would be.

Author: It's not, though; `KUBE-IPTABLES-HINT` is created by kubelet, regardless of the kube-proxy mode (or even if you're using a third-party proxy implementation). It's not "HINT that KUBErnetes is using IPTABLES", it's "HINT provided by KUBErnetes on how to use IPTABLES".

  - [Failure mode brief description]
    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

- Initial proposal: 2023-02-01

## Drawbacks

Adding a new officially-supported kube-proxy implementation implies
more work for SIG Network (especially if we are not able to deprecate
either of the existing backends soon).

Replacing the default kube-proxy implementation will affect many
users.

However, doing nothing would result in a situation where, eventually,
many users would be unable to use the default proxy implementation.

## Alternatives

> 

Reviewer 3: We could list ”reintroduce userland kube-proxy” - perhaps with static nftables config, perhaps with `NFQUEUE` - and then explain why this isn't going to perform how we want.

Author: I don't know... that just seems like an obviously terrible idea that we don't even have to humor :joy:

Reviewer 3: Not actually obvious to me, but feel free to mark this resolved anyway.

    kube-proxy container image"
  - "clearly not ready for defaulting"
  - "complications ... with IPVS kernel modules missing or disabled on
    user nodes"
  - "we are still lacking tests"
  - "still does not completely align with what [we] support in
    iptables mode"
  - "iptables works and people are familiar with it"
  - "[not sure that it was ever intended for IPVS to be the default]"

Additionally, the kernel IPVS APIs alone do not provide enough
functionality to fully implement Kubernetes services, and so the
`ipvs` backend also makes heavy use of the iptables API. Thus, if we
are worried about iptables deprecation, then in order to switch to
using `ipvs` as the default mode, we would have to port the iptables
parts of it to use nftables anyway. But at that point, there would be
little excuse for using IPVS for the core load-balancing part,
particularly given that IPVS, like iptables, is no longer an
actively-developed technology.

[kubeadm #817]: https://github.com/kubernetes/kubeadm/issues/817
[not sure that it was ever intended for IPVS to be the default]: https://en.wikipedia.org/wiki/The_Fox_and_the_Grapes

> ### The `nf_tables` mode of `/sbin/iptables` will not save us

Reviewer 4: ahhhh. i didnt know this thanks for clarifying this !!!!!!!

In 2018, with the 1.8.0 release of the iptables client binaries, a new
mode was added to the binaries, to allow them to use the nftables API
in the kernel rather than the legacy iptables API, while still
preserving the "API" of the original iptables binaries. As of 2022,
most Linux distributions now use this mode, so the legacy iptables

> kernel API is mostly dead.

Reviewer 5: This is somewhat adressed in comments (not the KEP though) further back:

> Yeah... so as I was explaining to Antonio, the normal way to do things in nftables is that everyone creates their own tables and doesn't interfere with anyone else's. In your table you create your own chains that connect to the standard hooks (forward, prerouting, etc) with particular priority values. So, if we document the hooks and priority levels where we do particular things, then people ought to be able to create their own rules in their own tables that run either before or after our rules as needed, just by using higher or lower priority values.

https://github.com/kubernetes/enhancements/pull/3824#issuecomment-1418010527

> kernel API is mostly dead.

Author: I think what you're asking about is covered by this in the Goals:

```
    - Document specific details of the nftables implementation that we
      want to consider as "API". (eg, documenting the `priority` of
      certain chains so that administrators or other components can
      reliably create other chains that act either before or after
      kube-proxy's.)
```

with nftables.

### The `iptables` mode of kube-proxy has grown crufty

Because `iptables` is the default kube-proxy mode, it is subject to
strong backward-compatibility constraints which mean that certain
"features" that are now considered to be bad ideas cannot be removed
because they might break some existing users. A few examples:

  - It allows NodePort services to be accessed on `localhost`, which
    requires it to set a sysctl to a value that may introduce security
    holes on the system. More generally, it defaults to having
    NodePort services be accessible on _all_ node IPs, when most users
    would probably prefer them to be more restricted.

  - It implements the `LoadBalancerSourceRanges` feature for traffic
    addressed directly to LoadBalancer IPs, but not for traffic
    redirected to a NodePort by an external LoadBalancer.

  - Some new functionality only works correctly if the administrator
    passes certain command-line options to kube-proxy (eg,
    `--cluster-cidr`), but we cannot make those options be mandatory,
    since that would break old clusters that aren't passing them.

A new kube-proxy mode, which existing users would have to explicitly opt

> into, could revisit these and other decisions.

Author: [Tim suggested earlier](https://github.com/kubernetes/enhancements/pull/3824#discussion_r1096604526) that we should think about still supporting it, if we can support it without requiring the `route_localnet` sysctl to be set.

For the most part, it ought to be possible for people to change their code to connect to `${NODE_IP}:${NODE_PORT}` rather than `127.0.0.1:${NODE_PORT}`. In discussions elsewhere, I believe we've only come up with one known use case where you can't do that, which is that docker will let you connect to a registry without https if it's on `127.0.0.1`, so if you use a localhost nodeport to connect to a docker registry pod, you don't need to generate an https cert.

> into, could revisit these and other decisions.

Reviewer 5: Though there's also a chicken and egg issue of knowing a nodeIP? localhost also works for pre-configuring this sort of thing.

### We will hopefully be able to trade 2 supported backends for 1

Right now SIG Network is supporting both the `iptables` and `ipvs`
backends of kube-proxy, and does not feel like it can ditch `ipvs`
because of perceived performance issues with `iptables`. If we create a new
backend which is as functional and non-buggy as `iptables` but as
performant as `ipvs`, then we could (eventually) deprecate both of the

> existing backends and only have one backend to support in the future.

Reviewer 3: ```suggestion
existing backends and only have one Linux backend to support in the future.
```

      rather than just logging warnings about missing functionality.

    - Consider various fixes to legacy `iptables` mode behavior.

        - In particular, ensure that we do not set the
          `route_localnet` sysctl by default, either because we don't
          support localhost nodeports by default, or because we
          implement them in some way that doesn't require the sysctl.

    - Document specific details of the nftables implementation that we
      want to consider as "API". (eg, documenting the `priority` of
      certain chains so that administrators or other components can
      reliably create other chains that act either before or after
      kube-proxy's.)

- Come up with at least a vague plan to eventually make `nftables` the
  default backend.

- Decide whether we can/should deprecate or even remove the `iptables`
  and/or `ipvs` backends. (Perhaps they can be pushed out of tree, a
  la `cri-dockerd`.)

- Help with the clean-up and refactoring of the kube-proxy "library"
  code.

### Non-Goals

- Falling into the same traps as the `ipvs` backend, to the extent
  that we can identify what those traps were.

> - Removing the iptables `KUBE-IPTABLES-HINT` chain from kubelet; that

Author: Right. On RHEL 10, kubelet will try to create the hint, fail, and log something reasonable. But on other platforms where people might still be using iptables-based components, we'll still want kubelet to create the hint so those components can use it.

ignore our table, so we should not need to worry about watching for
other people deleting our rules like we have to in the `iptables`
backend.

(In theory, instead of creating one table each in the `ip` and `ip6`
families, we could create a single table in the `inet` family and put
both IPv4 and IPv6 chains/rules there. However, this wouldn't really
result in much simplification, because we would still need separate
sets/maps to match IPv4 addresses and IPv6 addresses. (There is no
data type that can store/match either an IPv4 address or an IPv6
address.) Furthermore, because of how Kubernetes Services evolved in
parallel with the existing kube-proxy implementation, we have ended up
with a dual-stack Service semantics that is most easily implemented by
handling IPv4 and IPv6 completely separately anyway.)

#### Communicating with the kernel nftables subsystem

At least initially, we will use the `nft` command-line tool to read
and write rules, much like how we use command-line tools in the
`iptables` and `ipvs` backends. However, the `nft` tool is mostly just
a thin wrapper around `libnftables`, and it would be possible to use

> that directly instead in the future, given a cgo wrapper.

Author: Huh. I didn't realize we produced binaries that people ran outside of containers that we also produced.

Anyway, we don't _have to_ use the library. I was mostly pointing out the "thin wrapper" part and suggesting that even if our golang API has obvious nft-command-line-isms in it, that wouldn't prevent us from switching to using libnftables later.

> that directly instead in the future, given a cgo wrapper.

Reviewer 5: We compile almost everything with CGO_ENABLED=0 to simplify cross compiling and distribution but not kubelet and I think we get by there mostly because the build env is Debian stable so most distros have >= libc etc.

Fair, we also don't have to not use it, but non-static binaries do make build & distribution trickier.

When reading data from the kernel (`nft list ...`), `nft` outputs the
data in a nested "object" form:

```
table ip kube_proxy {
  comment "Kubernetes service proxying rules";

  chain services {
    ip daddr . ip protocol . th dport vmap @service_ips
  }
}
```

(This is the "native" nftables syntax, but the tools also support a

> JSON syntax that may be easier for us to work with...)

Author: Yes, although we won't have to do a lot of actual parsing. (For comparison, the iptables backend never parses actual iptables rules. The closest it comes is that it parses the _chain names_ out of `iptables-save` so it can figure out if there are stale chains that should have been deleted earlier but weren't.)

    around in a way that bypassed the kernel checks that otherwise
    make this not work?

  - We could resurrect enough of the userspace proxy to use it just
    for localhost NodePort connections... :vomiting_face:

      - (Alternatively: having a listening port on localhost, but then
        after accepting a connection, using eBPF to splice it directly
        to the destination server socket.)

<<[/UNRESOLVED]>>
```

```
<<[UNRESOLVED localhost nodeport use cases ]>>

It's not clear what the use cases for localhost NodePorts are. The
initial implementation of NodePorts in the iptables proxy
(https://github.com/kubernetes/kubernetes/pull/12764) set
`route_localnet`, so localhost nodeports seem to have been considered
an expected feature already at that point.

  - In https://github.com/kubernetes/kubernetes/issues/90236#issuecomment-622142806
    @thockin commented that one use case is connecting to a docker
    registry without needing to set up certificates; docker
    special-cases `localhost` specifically as not needing TLS. (@uablrek
    also mentions in https://github.com/kubernetes/kubernetes/pull/92938#issuecomment-742985382

>     that this is "the only use-case I have heard".)

Reviewer 5: Not just for certificates: it's useful because we want the container runtime on the nodes to pull from it.

So I run a nodePort service pointed at the registry running inside the cluster, and then I pull the image on any node with localhost:port

The container runtime itself can't resolve in-cluster DNS so using localhost nodeports solves this pretty elegantly at the moment.

We use this in SIG Testing for development type purposes, where the image never needs to be pushed to a remote host.

I don't know how widespread this is elsewhere, and I'm not saying this should be a prioritized use-case, but this seems difficult to reasonably implement without localhost nodeport. I suppose we could run a hostPort proxy daemonset ... 

EX: https://kind.sigs.k8s.io/docs/user/local-registry/ (where certificates aren't really the concern, though that is also handy)

Author: If the only thing you need is "has a statically-known IP" then there are other options. By default you can reach a NodePort service on _any_ local IP, so if you know of another IP that is present (eg, `172.17.0.1` for `docker0`), you can access the service on that IP instead. (It looks like `10.244.0.1` should work in the default kind configuration?)

Alternatively, you could configure the service with an `externalIPs` and use that IP to access the service. (Ideally you'd pick an IP from a private range that wasn't already in use, but failing that you could just do `externalIPs: ["1.2.3.4"]` or whatever...)

Reviewer 5: VIPs implemented by kube-proxy won't work for "in cluster registrY" use-case AFAIK because the container runtime on the node has to access it (containerd/cri-o/...) as opposed to a non-hostNetwork pod (?)

NodePort being bound to loopback however is reachable from the container runtime on each node in the host network.

Author: (as discussed in the linked issue, the container runtime _does_ have access to the service CIDR)

anyway, this section has been rewritten since these comments

rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

<!--
Describe manual testing that was done and the outcomes.
Longer term, we may want to require automated upgrade/rollback tests, but we
are missing a bunch of machinery and tooling and can't do that now.
-->

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

<!--
Even if applying deprecation policies, they may still surprise some users.
-->

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### How can an operator determine if the feature is in use by workloads?

The feature is used by the cluster as a whole, and the operator would

> know that it was in use from looking at the cluster configuration.

Reviewer 3: Can an cluster operator use one proxy mode on one set of nodes and a different proxy mode on another set of nodes? (I think so, but have never tried it).

Author: Yes. I even talk about that somewhere else (that it's possible to switch from iptables to nftables in a running cluster, because nodes don't care what mode other nodes are running so you don't have to worry about skew).

Reviewer 3: Given that, it's useful to reword - but that fix can happen after the KEP merges as provisional.

entirely clear why... [kubeadm #817], "Track when we can enable the
ipvs mode for the kube-proxy by default" is perhaps a good snapshot of
the initial excitement followed by growing disillusionment with the
`ipvs` mode:

  - "a few issues ... re: the version of iptables/ipset shipped in the
    kube-proxy container image"
  - "clearly not ready for defaulting"
  - "complications ... with IPVS kernel modules missing or disabled on
    user nodes"
  - "we are still lacking tests"
  - "still does not completely align with what [we] support in
    iptables mode"
  - "iptables works and people are familiar with it"
  - "[not sure that it was ever intended for IPVS to be the default]"

Additionally, the kernel IPVS APIs alone do not provide enough
functionality to fully implement Kubernetes services, and so the
`ipvs` backend also makes heavy use of the iptables API. Thus, if we
are worried about iptables deprecation, then in order to switch to
using `ipvs` as the default mode, we would have to port the iptables
parts of it to use nftables anyway. But at that point, there would be
little excuse for using IPVS for the core load-balancing part,

> particularly given that IPVS, like iptables, is no longer an

Reviewer 4: *question*
ahhh  I didnt know ipvs wasnt actively developed anymore.  cc @Damans227 .....

....  should we deprecate ipvs along the way? if we just whittle kube proxy down to one, NFT impl, that would make it super easy to refactor in the future? (maybe its a separate KEP, but , if were going all in on NFT, might as well deprecate iptables and IPVS together right) ?

Author: > Maybe linking to some link that represents the IPVS decline would be useful

Yeah, the problem is that the decline is more just of the "everyone stops paying attention to it" sort rather than the "formally deprecated" sort. There's just not a lot going on with that code, as compared to nftables (or eBPF). If you look at https://github.com/torvalds/linux/commits/master/net/netfilter/ipvs, there's a lot of "update the ipvs code to reflect this API change made elsewhere" type stuff, and a little bit of bugfixing (eg [this two-line fix](https://github.com/torvalds/linux/commit/d7fce52fdf96663ddc2eb21afecff3775588612a) that merged two weeks ago), but not really any new feature development.

> should we deprecate ipvs along the way?

That's an open question. As noted in the Motivation section, _this_ KEP is just about adding nftables (which will certainly result in us having to support all three of iptables, ipvs, and nftables for some amount of time). We'll have to continue to think about what comes next.

> IMo we could delete the windows proxy along the way ... if theres sufficient buy in from the community

That would actually help out some with the refactoring stuff that's going on in parallel with this KEP. Eg, as noted in https://github.com/kubernetes/kubernetes/pull/118464, there are simplifications we could make to the `pkg/proxy` code if we only had to support iptables/ipvs, but the winkernel backend does things differently enough that it makes things harder to refactor (especially since I can't even test the winkernel proxy easily).

OTOH, we already keep screwing up on "generic" vs "Linux-only" (https://github.com/kubernetes/kubernetes/issues/117909), and moving the Windows code out of tree would make that worse.

Maybe we should put both the windows proxy _and_ the nftables proxy out-of-tree...

Reviewer 4: > Maybe we should put both the windows proxy and the nftables proxy out-of-tree...
Is that still open for debate? that would be nice and sorta help force the issue of making the kube proxy easier for people to understand, bc you'd have a reference, standalone impl w/o less  generic boiler plate.

Reviewer: Since most of existing kube-proxy problems config versions, flags, Windows... require breaking compatibility (yeah, alpha means alpha, but this was alpha for years and will cause a lot of disruption across a lot of projects). Being able to start from scratch is worth considering

Reviewer: though there are at least 2 other previous nftables implementations that I know of , and they went nowhere ... I like Dan's approach in the KEP,  the main driver is that `iptables` is going to be removed and become unmaintained

Reviewer: he also explains this in L263, we can not become default this new mode if we break compatibility, users deployments will stop working

addressed by testing, and by a slow, optional, rollout of the new proxy
mode.

The `nftables` mode should not pose any new security issues relative
to the `iptables` mode.

## Design Details

### High level design

At a high level, the new mode should have the same architecture as the
existing modes; it will use the service/endpoint-tracking code in
`k8s.io/kubernetes/pkg/proxy` to watch for changes, and update rules
in the kernel accordingly.

### Low level design

Some details will be figured out as we implement it. We may start with
an implementation that is architecturally closer to the `iptables`
mode, and then rewrite it to take advantage of additional nftables
features over time.

#### Tables

Unlike iptables, nftables does not have any reserved/default tables or
chains (eg, `nat`, `PREROUTING`). Users are expected to create their
own tables and chains for their own purposes. An nftables table can
only contain rules for a single "family" (`ip` (v4), `ip6`, `inet`
(both IPv4 and IPv6), `arp`, `bridge`, or `netdev`), but unlike in
iptables, you can have both "filter"-type chains and "NAT"-type chains
in the same table.

> So, we will create a single `kube_proxy` table in the `ip` family, and

Author: In this particular case, no, because no packet would ever see both tables anyway; regardless of whether we have a single `inet kube-proxy` table or one `ip kube-proxy` table and one `ip6 kube-proxy` table, any given IPv4 or IPv6 packet will pass through exactly one `kube-proxy` table.

add rule ip kube_proxy services ip daddr . ip protocol . th dport vmap @service_ips
```

The "object" form is more logical and easy to understand, but the
"command" form is better for dynamic usage. In particular, it allows
you to add and remove individual chains, rules, map/set elements, etc,
without needing to also include the chains/rules/elements that you are
not modifying.

The examples below all show the "object" form of data, but it should
be understood that these are examples of what would be seen in `nft
list` output after kube-proxy creates the rules (with additional
`#`-preceded comments added to help the KEP reader), not examples of
the data we will actually be passing to `nft`.

The examples below are also all IPv4-specific, for simplicity. When
actually writing out rules for nft, we will need to switch between,
e.g., "`ip daddr`" and "`ip6 daddr`" appropriately, to match an IPv4
or IPv6 destination address. This will actually be fairly simple
because the `nft` command lets you create "variables" (really
constants) and substitute their values into the rules. Thus, we can
just always have the rule-generating code write "`$IP daddr`", and

> then pass either "`-D IP=ip`" or "`-D IP=ip6`" to `nft` to fix it up.)

Reviewer: I prefer remove as much as possible any dependency on the userspace tool `nft` and do all this in golang ... is easier to understand and removes any chance of hitting bugs on these features

Author: So, as I've been working on this, I've discovered that the situation with the nftables API is kind of more ugly than I'd thought. The underlying netlink API (which is not really documented anywhere) is very very different from the text-based command-line API (which is documented in the man page, the wiki, lots of random third-party blog posts, etc). There is also a JSON API, which is mostly a mapping of the netlink API (meaning that it is also poorly documented and hard to translate to and from the textual command-line API).

So we have to choose between:
1. use the API that 99% of documentation and examples on the web use
2. use the API that is closer to the bits and less prone to hitting random bugs

I feel like for maintainability purposes, we really have to go with 1.

Reviewer: I was referring more to the use of "special" feature as variable replacement ... I think that 1. is ok if we achieve the same degree that we got in iptables

(Also, most of the examples below have not actually been tested and
may have syntax errors. Caveat lector.)

#### Versioning and compatibility

Since nftables is subject to much more development than iptables has
been recently, we will need to pay more attention to kernel and tool
versions.

The `nft` command has a `--check` option which can be used to check if
a command could be run successfully; it parses the input, and then
(assuming success), uploads the data to the kernel and asks the kernel
to check it (but not actually act on it) as well. Thus, with a few

> `nft --check` runs at startup we should be able to confirm what

Reviewer: great, this is much better than existing tooling that checks if flag exist or minor versions is greater than ...

features are known to both the tooling and the kernel.

It is not yet clear what the minimum kernel or `nft` command-line
versions needed by the `nftables` backend will be. The newest feature
used in the examples below was added in Linux 5.6, released in March
2020 (though they could be rewritten to not need that feature).

It is possible some users will not be able to upgrade from the
`iptables` and `ipvs` backends to `nftables`. (Certainly the
`nftables` backend will not support RHEL 7, which some people are
still using Kubernetes with.)

#### NAT rules

##### General Service dispatch

For ClusterIP and external IP services, we will use an nftables
"verdict map" to store the logic about where to dispatch traffic,
based on destination IP, protocol, and port. We will then need only a

> single actual rule to apply the verdict map to all inbound traffic.

Reviewer: let me see if I understand, right now the Service. logic is O(N) , so you iterate every single rules, does this means that this verdict map will make this O(1)?

Author: Yes, O(1) on the "data plane" side. On the "control plane" side, adding/removing elements from a vmap isn't O(1) because there's sanity-checking (eg making sure that you haven't created a loop with `jump`/`goto` rules). It's not clear how big that cost is at large-k8s-cluster scale (though _at worst_ it would still be doing less work than iptables with `MinimizeIPTablesRestore`)

Reviewer: will be interesting to see the output of the tests 👍

  numgen random mod 2 vmap {
    0 : goto sep_UKSFD7AGPMPPLUHC,
    1 : goto sep_C6EBXVWJJZMIWKLZ
  }
}

# Example per-endpoint chain
chain sep_UKSFD7AGPMPPLUHC {
  # masquerade hairpin traffic
  ip saddr 10.180.0.4 jump mark_for_masquerade

  # send to selected endpoint
  dnat to 10.180.0.4:8000
}
```

##### Masquerading

The example rules above include

```
  ip saddr 10.180.0.4 jump mark_for_masquerade
```

to masquerade hairpin traffic, as in the `iptables` proxier. This
assumes the existence of a `mark_for_masquerade` chain, not shown.

nftables has the same constraints on DNAT and masquerading as iptables
does; you can only DNAT from the "prerouting" stage and you can only
masquerade from the "postrouting" stage. Thus, as with `iptables`, the
`nftables` proxy will have to handle DNAT and masquerading at separate
times. One possibility would be to simply copy the existing logic from
the `iptables` proxy, using the packet mark to communicate from the
prerouting chains to the postrouting ones.

However, it should be possible to do this in nftables without using
the mark or any other externally-visible state; we can just create an

> nftables `set`, and use that to communicate information between the

Reviewer: I would be interesting in knowing the performance implications of using this vs a mark

chains. Something like:

```
# Set of 5-tuples of connections that need masquerading
set need_masquerade {
  type ipv4_addr . inet_service . ipv4_addr . inet_service . inet_proto;
  flags timeout ; timeout 5s ;
}

chain mark_for_masquerade {
  update @need_masquerade { ip saddr . th sport . ip daddr . th dport . ip protocol }
}

chain postrouting_do_masquerade {
  # We use "ct original ip daddr" and "ct original proto-dst" here
  # since the packet may have been DNATted by this point.

  ip saddr . th sport . ct original ip daddr . ct original proto-dst . ip protocol @need_masquerade masquerade
}
```

This is not yet tested, but some kernel nftables developers have
confirmed that it ought to work.

##### Session affinity

Session affinity can be done in roughly the same way as in the
`iptables` proxy, just using the more general nftables "set" framework
rather than the affinity-specific version of sets provided by the
iptables `recent` module. In fact, since nftables allows arbitrary set
keys, we can optimize relative to `iptables`, and only have a single

> affinity set per service, rather than one per endpoint. (And we also

Reviewer: yeah, every certain period someone demands this feature

Author: yes, this is following up on the discussion when we removed SessionAffinity timeout behavior from Conformance (https://github.com/kubernetes/kubernetes/pull/112806)

}

chain filter {
  ...
  ip daddr . ip protocol . th dport @firewall jump firewall_check
  ...
}

chain firewall_check {
  ip daddr . ip protocol . th dport . ip saddr @firewall_allow return
  drop
}
```

Where, eg, adding a Service with LoadBalancer IP `10.1.2.3`, port
`80`, and source ranges `["192.168.0.3/32", "192.168.1.0/24"]` would
result in:

```
add element ip kube_proxy firewall { 10.1.2.3 . tcp . 80 }
add element ip kube_proxy firewall { 10.1.2.3 . tcp . 80 }
add element ip kube_proxy firewall_allow { 10.1.2.3 . tcp . 80 . 192.168.0.3/32 }
add element ip kube_proxy firewall_allow { 10.1.2.3 . tcp . 80 . 192.168.1.0/24 }
```

##### Forcing traffic on `HealthCheckNodePorts` to be accepted

The `iptables` mode adds rules to ensure that traffic to NodePort
services' health check ports is allowed through the firewall. eg:

```
-A KUBE-NODEPORTS -m comment --comment "ns2/svc2:p80 health check node port" -m tcp -p tcp --dport 30000 -j ACCEPT
```

(There are also rules to accept any traffic that has already been
tagged by conntrack.)

This cannot be done reliably in nftables; the `accept` and `drop`
rules work differently than they do in iptables, and so if there is a

> firewall that would drop traffic to that port, then there is no

Reviewer: what does this mean exactly? in terms of tables and rules

Author: `-j ACCEPT` in iptables and `accept` in nftables mean "stop processing further rules _in the current table_". In the case of iptables, since you mostly only call `-j ACCEPT`/`-j DROP` from the `filter` table, that means that if you `ACCEPT` something early on, it doesn't matter if someone else has a rule that would `DROP` it later.

But with nftables, there can be lots of tables that have chains attached to the `filter` hook, so if you call `accept` in one table, the packet could still hit a `drop` rule later in _another_ table. You can only reliably use `accept` to prevent a packet from being dropped by _your own_ rules, not by someone else's rules.

guaranteed way to "sneak behind its back" like you can in iptables; we
would need to actually properly configure _that firewall_ to accept
the packets.

> However, these sorts of rules are somewhat legacy anyway; they work

Reviewer 4: i think we have e2es for these in 
```
                                   
2708   ginkgo.It("should fail health check node port if there are only terminating endpoints [Feature:ProxyTerminatingEndpoints]", func(ctx context.Context) {
2709     // windows kube-proxy does not support this feature yet                                                                                                                                                               
2710     e2eskipper.SkipIfNodeOSDistroIs("windows")                                  
2711                                                                                 
2712     nodes, err := e2enode.GetBoundedReadySchedulableNodes(ctx, cs, 2)           
2713     framework.ExpectNoError(err)                                                
2714     nodeCounts := len(nodes.Items)                                              
2715     if nodeCounts < 2 {                                                         
2716       e2eskipper.Skipf("The test requires at least two ready nodes on %s, but found %v", framework.TestContext.Provider, nodeCounts)
2717     }                                                                           
2718     node0 := nodes.Items[0]                         
```

does that mean that those tests will break in clusters where this *used to* work bc of the god-power of iptables?

Author: That is testing that the HealthCheckNodePort answers "healthy"/"not-healthy" correctly.

The thing that I was talking about is that with the iptables proxy, someone could configure their iptables-based firewall to drop all packets, but HealthCheckNodePorts would still work, because iptables kube-proxy adds an explicit `ACCEPT` rule to override any firewall `DROP` rules.

With nftables, that won't work; if the firewall table says to drop all packets, you can't override that from the `kube-proxy` table. The administrator would have to reconfigure their firewall to say "drop all packets except for packets in the NodePort range" instead. (Like they already have to do if they are configuring a cloud-network-level firewall rather than a node-level iptables firewall.)

Reviewer 4: yup so my q was 
- if admin is dropping those packets
- and upgrade from iptables to the -> nft proxy 

... will this (or other e2es) start breaking ?

Author: Ah, that's not really a question about the e2e tests; the important question is "if the admin is dropping those packets, and they switch from iptables to nftables, _will HealthCheckNodePorts start breaking_?" And the currently-planned answer to that is "yes", as explained in the section you're commenting on. The plan is for kube-proxy to stop trying to work around incorrect firewall rules, and instead require that the firewall rules actually be correct.

For this and other reasons, as discussed in the KEP, the nftables mode would not be a drop-in replacement for iptables. It may require manual configuration changes.

convenient) to just rewrite things to use the node IP rather than
localhost (or to use a ClusterIP rather than a NodePort). Indeed,
since localhost NodePorts do not work with `ipvs` mode or with IPv6,
many places that used to use NodePorts on `127.0.0.1` have already
been rewritten to not do so (eg [contiv/vpp#1434]).

So:

  - There is no way to make IPv6 localhost NodePorts work with a
    NAT-based solution.

  - The way to make IPv4 localhost NodePorts work with NAT introduces
    a security hole, and we don't necessarily have a fully-generic way
    to mitigate it.

  - The only commonly-argued-for use case for the feature involves
    deploying a service in a configuration which its own documentation
    describes as insecure and "only appropriate for testing".

      - The use case in question works by default against cri-dockerd
        but not against containerd or cri-o with their default
        configurations.

      - cri-dockerd, containerd, and cri-o all allow additional
        "insecure registry" IPs/CIDRs to be configured, so an
        administrator could configure them to allow non-TLS image
        pulling against a ClusterIP.

> Given this, I think we should not try to support localhost NodePorts

Reviewer: tend to agree, anyway, since this is going to take several releases we can always reevaluate or leave it as a. new option, right now we are skipping it based on a flag value containing the localhost address or no

Reviewer 2: My only ask would be to ensure the differences like this one are clearly articulated in the docs, so users are aware about the change.

accepting NodePort connections on all local IPs, which has effects
varying from intended-but-unexpected ("why can people connect to
NodePort services from the management network?") to clearly-just-wrong
("why can people connect to NodePort services on LoadBalancer IPs?")

The nftables proxy should default to only opening NodePorts on a
single interface, probably the interface with the default route by
default. (Ideally, you really want it to accept NodePorts on the
interface that holds the route to the cloud load balancers, but we
don't necessarily know what that is ahead of time.) Admins can use
`--nodeport-addresses` to override this.

#### Behavior of service IPs

```
<<[UNRESOLVED unused service IP ports ]>>

@thockin has suggested that service IPs should reject connections on
ports they aren't using. (This would most easily be implemented by
adding a `--service-cidr` flag to kube-proxy so we could just "reject
everything else", but even without that we could at least reject
connections on inactive ports of active service IPs.)

<<[/UNRESOLVED]>>
```

```
<<[UNRESOLVED service IP pings ]>>

Users sometimes get confused by the fact that service IPs do not

> respond to ICMP pings, and perhaps this is something we could change.

Reviewer: I do not think we should respond to ping, Services are about IP:port:protocol, ICMP does not have to be the tool to resolve the problem and is common to filter ICMP ... people should use the right tool to solve the problem, that in this case is TCP or UDP connectivity,

Reviewer 3: We could _document_ that we don't guarantee to.

Reviewer 6: Supporting ping is nice and friendly, and I don't see any reason NOT to really, but I also don't see a strong reason in favor.  One could argue that it's an info-leak but DNS is already in play :)

Reviewer 7: A problem with ping is that it has no ports. So, ping may go to any service that have the address (external), or to any protocol endpoint, tcp, udp or sctp (both external and clusterip). IMHO ping is useful in _most_ cases, and users expects it, but the corner cases are hard to explain.

Author: I was only talking about ping to ClusterIPs, and in that case, ClusterIPs map one-to-one to Services (or, well, two-to-one for dual-stack), so port wouldn't really matter. The idea would be either (a) pings to a ClusterIP get DNATted to an endpoint IP (and then handled normally by the kernel), or (b) pings to a ClusterIP get replied to automatically by nftables rules if the ClusterIP is in use, or (c) pings to a ClusterIP get replied to automatically by nftables rules if the ClusterIP is in use _and_ has active endpoints. Depending on what semantics we wanted.

Reviewer: there is this an edge case that one Service with mutiple ports can have different backends, so the same Service ClusterIP can map to different pods depending on the port ... the ping will be confusing because port X can be down and port Y working, and the ping will be intermittently failing

<<[/UNRESOLVED]>>
```

#### Integrating admin/debug/third-party rules with kube-proxy's nftables rules

Administrators sometimes want to add rules to log or drop certain
packets. Kube-proxy makes this difficult because it is constantly
rewriting its rules, making it likely that admin-added rules will be
deleted shortly after being added.

Likewise, external components (eg, NetworkPolicy implementations) may
want to write rules that integrate with kube-proxy's rules in
well-defined ways.

NFTables makes this easier than iptables in some ways, because each
application is expected to create their own table, and not interfere

> with anyone else's tables. If we document the `priority` values we use

Reviewer: yeah, we need to create this and we know that is going to become an API 😄

to connect to each nftables hook, then admins and third party
developers should be able to reliably process packets before or after
kube-proxy, without needing to modify kube-proxy's chains/rules.

In cases where administrators want to insert rules into the middle of
particular service or endpoint chains, it is hard to avoid
accidentally overwriting them (at least when the services/endpoints
actually change). The new `MinimizeIPTablesRestore` behavior in
iptables makes things a little better since we no longer rewrite

> service/endpoint chains that didn't change, so admin-added rules will

Reviewer: and what happens if this admin rules break the service?
it feels wrong allowing to set random rules and the set should be reconciled to match the Service IMHO

Reviewer: got it, sounds good

- Complete features A, B, C
- Additional tests are in Testgrid and linked in KEP

#### GA

- N examples of real-world usage
- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

### Upgrade / Downgrade Strategy

The new mode should not introduce any upgrade/downgrade problems,
excepting that you can't downgrade or feature-disable a cluster using
the new kube-proxy mode without switching it back to `iptables` or

> `ipvs` first.

Reviewer: we have this `kube-proxy --cleanup `flag for this, right?

Reviewer: ok, is in the next paragraph

###### How can an operator determine if the feature is in use by workloads?

The operator is the one who would enable the feature, and they would
know it is in use by looking at the kube-proxy configuration.

###### How can someone using this feature know that it is working for their instance?

- [X] Other (treat as last resort)
  - Details: If Services still work then the feature is working

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

<!--
This is your opportunity to define what "normal" quality of service looks like
for a feature.

It's impossible to provide comprehensive guidance, but at the very
high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [X] Metrics

>   - Metric names:

Reviewer: we should have something similar to existing

> sync_proxy_rules_iptables_total

Author: It is really unclear what metrics we are going to want... a metric that was actually "total number of rules" might not be interesting because so much of the service processing is offloaded to set/map lookups. We could do "total combined number of rules and set/map elements", and that would tell you ... _something_, but it's not clear _what_. Especially, it's not clear that performance will have any reliable correlation to either "number of rules" or "number of rules+elements".

So, yes, we want metrics, but I'm not sure what they're going to be...

e.g., "`ip daddr`" and "`ip6 daddr`" appropriately, to match an IPv4
or IPv6 destination address. This will actually be fairly simple
because the `nft` command lets you create "variables" (really
constants) and substitute their values into the rules. Thus, we can
just always have the rule-generating code write "`$IP daddr`", and
then pass either "`-D IP=ip`" or "`-D IP=ip6`" to `nft` to fix it up.)

(Also, most of the examples below have not actually been tested and
may have syntax errors. Caveat lector.)

`nft` also has a JSON mode which theoretically should be easier to use
(both in terms of parsing `nft` output and in terms of not having to
worry as much about the quoting of input). However, the JSON syntax of
individual chain rules is poorly documented, and often _extremely_
different from the non-JSON syntax. (In general, the only way to
reliably figure out what the JSON syntax is for a particular rule is
to get `nft` to translate it for you.) So specifying rules this way
would be harder to understand and maintain. For now, at least, the
plan is to always use the non-JSON syntax, at least for anything
involving rules or set/map elements.

> 

Reviewer 3: What stability guarantees are available for the non-JSON interface? Do we want to look these up before beta?

Author: Yeah, I wondered if I should update this section some more...

I have been talking with one of the nftables developers. The CLI API is considered "stable", but without strong ideas about exactly what that means (eg, right now the output has whitespace both before and after every kind of punctuation except commas, which have whitespace after but not before... Is that guaranteed? :man_shrugging:). He's going to bring it up at the [Netfilter Workshop](https://workshop.netfilter.org/2023/) in a few weeks.

I also realized while talking with him that we can probably get away with using the non-JSON API for the "write" commands (`add`, `delete`, `flush`, `insert`, `replace`), but use the JSON API for the "read" commands (`list`, `get`), because for kube-proxy's purposes, we never actually need to read back a `rule` we previously created, and for every type other than `rule`, we can parse the JSON responses correctly. (The JSON syntax for map/set elements is _annoying_, but not actually unusable.)

Reviewer 2: Did you ask about the `libnftables` API guarantees?

Author: The Netfilter Workshop was the last two days... no word yet on what came out of discussions. But I've already started implementing what I said in the comment above; we can use the JSON mode for things we need to parse the results of, because for our purposes we don't need to be able to parse back rules we previously created.

(or `-j ACCEPT` in iptables) is to end processing _of the current
table_. In iptables, this effectively guarantees that the packet is
accepted (since `-j ACCEPT` is mostly only used in the `filter`
table), but in nftables, it is still possible that someone would later
call `drop` on the packet from another table, causing it to be
dropped. There is no way to reliably "sneak behind the firewall's
back" like you can in iptables; if an nftables-based firewall is
dropping kube-proxy's packets, then you need to actually configure
_that firewall_ to accept them instead.

However, this firewall-bypassing behavior is somewhat legacy anyway;
the `iptables` proxy is able to bypass a _local_ firewall, but has no
ability to bypass a firewall implemented at the cloud network layer,
which is perhaps a more common configuration these days anyway.
Administrators using non-local firewalls are already required to
configure those firewalls correctly to allow Kubernetes traffic
through, and it is reasonable for us to just extend that requirement
to administrators using local firewalls as well.

Thus, the `nftables` backend will not attempt to replicate these

> `iptables`-backend rules.

Reviewer 3: Before GA, we could document how a local nftables packet filter can integrate with Kubernetes with `kube-proxy` in nftables mode. Would be nice.

Author: That's basically what the "#### Integrating admin/debug/third-party rules with kube-proxy's nftables rules" section is about (and it has an UNRESOLVED).

#### Future improvements

Further improvements are likely possible.

For example, it would be nice to not need a separate "hairpin" check for
every endpoint. There is no way to ask directly "does this packet have
the same source and destination IP?", but the proof-of-concept [kpng
nftables backend] does this instead:

```
set hairpin {
  type ipv4_addr . ipv4_addr;
  elements {
    10.180.0.4 . 10.180.0.4,
    10.180.0.5 . 10.180.0.5,
    ...
  }
}

chain ... {
  ...
  ip saddr . ip daddr @hairpin jump mark_for_masquerade
}
```

More efficiently, if nftables eventually got the ability to call eBPF
programs as part of rule processing (like iptables's `-m ebpf`) then
we could write a trivial eBPF program to check "source IP equals
destination IP" and then call that rather than needing the giant set
of redundant IPs.

If we do this, then we don't need the per-endpoint hairpin check
rules. If we could also get rid of the per-endpoint affinity-updating
rules, then we could get rid of the per-endpoint chains entirely,
since `dnat to ...` is an allowed vmap verdict:

```
chain svc_4SW47YFZTEDKD3PK {

>   # FIXME handle affinity somehow

Reviewer 3: Should we add an `UNRESOLVED` marker for this?

Author: That's in the "#### Future improvements" section, just pointing out some things that we _might_ do, possibly post-GA. (So UNRESOLVED doesn't seem right, since it's not something we have to resolve before GA.)

I suppose I could just drop that section... it's not really _necessary_.

### Upgrade / Downgrade Strategy

The new mode should not introduce any upgrade/downgrade problems,
excepting that you can't downgrade or feature-disable a cluster using
the new kube-proxy mode without switching it back to `iptables` or
`ipvs` first.

When rolling out or rolling back the feature, it should be safe to
enable the feature gate and change the configuration at the same time,
since nothing cares about the feature gate except for kube-proxy
itself. Likewise, it is expected to be safe to roll out the feature in
a live cluster, even though this will result in different proxy modes
running on different nodes, because Kubernetes service proxying is
defined in such a way that no node needs to be aware of the
implementation details of the service proxy implementation on any
other node.

(However, see the notes below in [Feature Enablement and
Rollback](#feature-enablement-and-rollback) about stale rule cleanup
when switching modes.)

### Version Skew Strategy

The feature is isolated to kube-proxy and does not introduce any API
changes, so the versions of other components do not matter.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

> <!--
> This section must be completed when targeting alpha to a release.
> -->

Reviewer 3: ```suggestion
```
It's filled in.

shared code again, and perhaps give us new ideas on how it can be
cleaned up, rationalized, and optimized.

[KEP-3786]: https://github.com/kubernetes/enhancements/issues/3786

### Goals

- Design and implement an `nftables` mode for kube-proxy.

    - Consider various fixes to legacy `iptables` mode behavior.

        - Do not enable the `route_localnet` sysctl.

        - Add a more restrictive startup mode to kube-proxy, which
          will error out if the configuration is invalid (e.g.,
          "`--detect-local-mode ClusterCIDR`" without specifying
          "`--cluster-cidr`") or incomplete (e.g.,
          partially-dual-stack but not fully-dual-stack).

        - (Possibly other changes discussed in this KEP.)

    - Document specific details of the nftables implementation that we
      want to consider as "API". (eg, documenting the `priority` of
      certain chains so that administrators or other components can
      reliably create other chains that act either before or after
      kube-proxy's.)

- Document incompatible changes between `iptables` mode and `nftables`
  mode (e.g. localhost NodePorts, firewall handling, etc).

> - Do some sort of performance testing comparing the `iptables`,

Reviewer 2: ```suggestion
- Do performance testing comparing the `iptables`,
```

  `ipvs`, and `nftables` backends in small, medium, and large
  clusters, comparing both the "control plane" aspects (time/CPU usage
  spent reprogramming rules) and "data plane" aspects (latency and
  throughput of packets to service IPs).

> - Come up with at least a vague plan to eventually make `nftables` the

Reviewer 2: Is it actually needed in this KEP? I thought that early on it was clearly stated that it'll be a subject of entirely new KEP.  Or maybe at least point out that it's optional goal.

  default backend.

> - Decide whether we can/should deprecate or even remove the `iptables`

Reviewer 2: Same as above, optional goal maybe?

Author: If we add the nftables backend and never manage to get rid of either of the other backends, then we've permanently made kube-proxy harder to maintain, which we don't want. So we shouldn't let ourselves get too far along in this process without figuring out what the long-term plan is.

Maybe that doesn't belong in the KEP...

that part of the test against `iptables` and `ipvs` proxies. (We are
currently assuming that we will not need an nftables version of this
test, since the problem of one component deleting another component's
rules should not exist with nftables.)

Though not directly related to kube-proxy, there are also other e2e
tests that use iptables which should eventually be ported to nftables;
notably, the ones using [`TestUnderTemporaryNetworkFailure`].407466

[It should recreate its iptables rules if they are deleted]: https://github.com/kubernetes/kubernetes/blob/v1.27.0/test/e2e/network/networking.go#L550
[`TestUnderTemporaryNetworkFailure`]: https://github.com/kubernetes/kubernetes/blob/v1.27.0-alpha.2/test/e2e/framework/network/utils.go#L1078

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

> - <test>: <link to test coverage>

Reviewer 2: Can you estimate what %coverage already existing tests provide, at least the backend-agnostic part? What additions should be considered to ensure the smooth transition between various beckends?

Reviewer 2: This is still open question

Author: I guess I don't really understand the question. We don't have any hard numbers for %coverage for e2e tests, and I added this paragraph trying to imply that the soft number is "big enough":

>     For the most part, we should not need to add any nftables-specific e2e
>     tests; the `nftables` backend's job is just to implement the Service
>     proxy API to the same specifications as the other backends do, so the
>     existing e2e tests already cover everything relevant.

Reviewer: there is no coverage in e2e as percentage, Dan's sentences LGTM from the testing perspective, hope is ok for PRR too

- The nftables mode has no bugs / regressions that would make us
  hesitate to recommend it.

- We have at least the start of a plan for the next steps (changing
  the default mode, deprecating the old backends, etc).

### Upgrade / Downgrade Strategy

The new mode should not introduce any upgrade/downgrade problems,
excepting that you can't downgrade or feature-disable a cluster using
the new kube-proxy mode without switching it back to `iptables` or
`ipvs` first.

When rolling out or rolling back the feature, it should be safe to
enable the feature gate and change the configuration at the same time,
since nothing cares about the feature gate except for kube-proxy
itself. Likewise, it is expected to be safe to roll out the feature in
a live cluster, even though this will result in different proxy modes
running on different nodes, because Kubernetes service proxying is
defined in such a way that no node needs to be aware of the
implementation details of the service proxy implementation on any
other node.

(However, see the notes below in [Feature Enablement and
Rollback](#feature-enablement-and-rollback) about stale rule cleanup
when switching modes.)

### Version Skew Strategy

> The feature is isolated to kube-proxy and does not introduce any API

Reviewer 2: Doesn't kube-proxy have versions? What happens if I use new setting with an older kube-proxy?

Author: That's not "Version Skew Strategy", that's "Upgrade / Downgrade Strategy".

(I don't answer that question directly there, it just says "you can't downgrade or feature-disable a cluster using the new kube-proxy mode without switching it back to `iptables` or `ipvs` first". As for what actually happens, 

`E0915 10:19:31.504389  447804 run.go:74] "command failed" err="failed validate: KubeProxyConfiguration.Mode.ProxyMode: Invalid value: \"nftables\": must be iptables, ipvs or blank (blank means the best-available proxy [currently iptables])"`

)

changes, so the versions of other components do not matter.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

The administrator must enable the feature gate to make the feature
available, and then must run kube-proxy with the
`--proxy-mode=nftables` flag.

Kube-proxy does not delete its rules on exit (to avoid service
interruptions when restarting/upgrading kube-proxy, or if it crashes).
This means that when switching between proxy modes, it is necessary
for the administrator to ensure that the rules created by the old
proxy mode get deleted. (Failure to do so may result in stale service
rules being left behind for an arbitrarily long time.) The simplest
way to do this is to reboot each node when switching from one proxy
mode to another, but it is also possible to run kube-proxy in "cleanup

> and exit" mode, eg:

Reviewer 2: This seems rather cumbersome? Can we somehow detect that proxy mode has changed and help user with the transition?

Author: kube-proxy used to do that and we decided it was terrible and removed the code...
I'll have to dig up the old discussions and figure out if those reasons still apply.

Author: oh, look, we wrote a KEP: [KEP-2448 Remove kube-proxy's automatic clean up logic](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2448-Remove-kube-proxy-automatic-clean-up-logic).

So the problem there was that the userspace, iptables, and ipvs proxy modes used some of the same chain names (especially, they all use `KUBE-SERVICES`), so trying to clean up the non-active modes would result in deleting some of the chains used by the active mode too, causing disruption until they were recreated.

This would not apply when switching to a backend based on an entirely different technology though, so we _could_ automatically clean up the iptables rules when switching to nftables mode, and vice versa.

`sync_proxy_rules_iptables_total`, which tells you the total number of
iptables rules kube-proxy has programmed. But the equivalent metric in
the `nftables` backend is not going to be as interesting, because many
of the things that are done with rules in the `iptables` backend will
be done with maps and sets in the `nftables` backend. Likewise, just
tallying "total number of rules and set/map elements" is not likely to
be useful, because the entire point of sets and maps is that they have
more-or-less **O(1)** behavior, so knowing the number of elements is
not going to give you much information about how well the system is
likely to be performing.

- [X] Metrics
  - Metric names:
      - ...
  - Components exposing the metric:
      - kube-proxy

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

<!--
Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
implementation difficulties, etc.).
-->

### Dependencies

<!--
This section must be completed when targeting beta to a release.
-->

###### Does this feature depend on any specific services running in the cluster?

> It may require a newer kernel than some current users have. It does

Reviewer 2: Can you explicitly define the min kernel level? You did mention that earlier in the doc, iirc.

Author: The examples in the KEP require kernel 5.6, but I discovered later that they require an `nft` binary which is more recent than that. I don't have a good sense of exactly what my current PoC requires (other than that it works on Fedora 36 but requires a few tweaks on RHEL 8).

So...

First, we need to figure out / decide what the required kernel and userspace versions will be. (My theory here is "the lowest common denominator of RHEL 8, and whatever version of Debian the official Kubernetes images are built against".)

Then we need to:
1. Make sure this makes it into the end user documentation.
2. Check this at runtime (with `nft --check` as discussed earlier)
3. Have clear internal documentation (and maybe assertions) to make sure people don't accidentally bump these requirements in the future.

Reviewer 8: That's fine - this is not required for Alpha and we will revisit it for Beta once we already will have a working version and will know much more about it.

Reviewer: related https://github.com/kubernetes/kubernetes/issues/116799

is not very well designed and is not something we would want to
support other people using in its current form.

Writing a new proxy backend will force us to look over all of this
shared code again, and perhaps give us new ideas on how it can be
cleaned up, rationalized, and optimized.

[KEP-3786]: https://github.com/kubernetes/enhancements/issues/3786

### Goals

- Design and implement an `nftables` mode for kube-proxy.

    - Consider various fixes to legacy `iptables` mode behavior.

        - Do not enable the `route_localnet` sysctl.

        - Add a more restrictive startup mode to kube-proxy, which
          will error out if the configuration is invalid (e.g.,
          "`--detect-local-mode ClusterCIDR`" without specifying
          "`--cluster-cidr`") or incomplete (e.g.,
          partially-dual-stack but not fully-dual-stack).

        - (Possibly other changes discussed in this KEP.)

    - Document specific details of the nftables implementation that we
      want to consider as "API". (eg, documenting the `priority` of
      certain chains so that administrators or other components can
      reliably create other chains that act either before or after

>       kube-proxy's.)

Reviewer 2: :+1:

chain reject_chain {
  reject
}
```

##### Dropping traffic rejected by `LoadBalancerSourceRanges`

The implementation of LoadBalancer source ranges will be similar to
the ipset-based implementation in the `ipvs` kube proxy: we use one
set to recognize "traffic that is subject to source ranges", and then
another to recognize "traffic that is _accepted_ by its service's
source ranges". Traffic which matches the first set but not the second
gets dropped:

```
set firewall {
  comment "destinations that are subject to LoadBalancerSourceRanges";
  type ipv4_addr . inet_proto . inet_service
}
set firewall_allow {
  comment "destination+sources that are allowed by LoadBalancerSourceRanges";
  type ipv4_addr . inet_proto . inet_service . ipv4_addr
}

chain filter {
  ...
  ip daddr . ip protocol . th dport @firewall jump firewall_check
  ...
}

chain firewall_check {
  ip daddr . ip protocol . th dport . ip saddr @firewall_allow return
  drop
}
```

Where, eg, adding a Service with LoadBalancer IP `10.1.2.3`, port
`80`, and source ranges `["192.168.0.3/32", "192.168.1.0/24"]` would
result in:

```
add element ip kube_proxy firewall { 10.1.2.3 . tcp . 80 }

> add element ip kube_proxy firewall { 10.1.2.3 . tcp . 80 }

Author: uh... we don't (and that would be a no-op). I must have cut+pasted wrong when duplicating the next line for the second source range.

     that might have been safe to change had we not set
     `route_localnet`, and which according to some reports [may block
     legitimate traffic] in some configurations).

  3. The Cilium approach: Intercept the connect(2) call with eBPF and
     rewrite the destination IP there, so that the network stack never
     actually sees a packet with destination `127.0.0.1` / `::1`. (As
     in the `userspace` kube-proxy case, this is not a special-case
     for localhost, it's just how Cilium does service proxying.)

  4. If you control the client, you can explicitly bind the socket to
     `127.0.0.1` / `::1` before connecting. (I'm not sure why this
     works since the packet still eventually gets routed off `lo`.) It
     doesn't seem to be possible to "spoof" this after the socket is
     created, though as with the previous case, you could do this by
     intercepting syscalls with eBPF.

In discussions about this feature, only one real use case has been
presented: it allows you to run a docker registry in a pod and then
have nodes use a NodePort service via `127.0.0.1` to access that
registry. Docker treats `127.0.0.1` as an "insecure registry" by

> default (though containerd and cri-o do not) and so does not require

Reviewer: I personally think that we should drop this hack of localhost nodeport and route_localnet, it is also promoting bad practices

> default (though containerd and cri-o do not) and so does not require

Reviewer 8: I guess you're all better positioned to make this decision. My only request for Beta will be that will not work the same way in nftables mode is easily to discover for operators of existing clusters (metrics).

> default (though containerd and cri-o do not) and so does not require

Reviewer 6: I agree that we can, cautiously, kill this.  We should do some outreach through usual channels and see if anyone screams.

application is expected to create their own table, and not interfere
with anyone else's tables. If we document the `priority` values we use
to connect to each nftables hook, then admins and third party
developers should be able to reliably process packets before or after
kube-proxy, without needing to modify kube-proxy's chains/rules.

In cases where administrators want to insert rules into the middle of
particular service or endpoint chains, we would have the same problem
that the `iptables` backend has, which is that it would be difficult
for us to avoid accidentally overwriting them when we update rules.
Additionally, we want to preserve our ability to redesign the rules
later to take better advantage of nftables features, which would be
impossible to do if we were officially allowing users to modify the
existing rules.

One possibility would be to add "admin override" vmaps that are
normally empty but which admins could add `jump`/`goto` rules to for
specific services to augment/bypass the normal service processing. It
probably makes sense to leave these out initially and see if people
actually do need them, or if creating rules in another table is
sufficient.

> 

Author: Kube-proxy operates as a set of black boxes. E.g., the `iptables` mode has more-or-less three boxes: one in the iptables `PREROUTING` chain that rewrites service IPs to endpoint IPs (somehow), one in the iptables `POSTROUTING` chain that does masquerading of packets that need it (somehow), and one in the `filter` table to handle various drop and reject rules (somehow). The `ipvs` backend has a slightly different set of black boxes (one of which is the kernel ipvs subsystem) and so packets don't behave in exactly the same way there, and likewise `userspace` was even more different. (And eBPF-based and OVS-based service proxies are _completely_ different.)

Technically, we do not actually support the existence of those black boxes as an "API", even though we know that various components like Calico require the boxes to work in certain ways, so we probably should do better there...

The instructions for [using kube-proxy in `ipvs` mode with Calico](https://docs.tigera.io/calico/latest/networking/configuring/use-ipvs) mention that "When Calico detects that kube-proxy is running in IPVS mode (during or after installation), IPVS support is automatically activated", suggesting that Calico needed tweaks to support the differences between `iptables` and `ipvs` mode. So, Calico may need tweaks for `nftables` mode too. Though probably less than it needed for `ipvs`; at the level of black boxes, there's not much difference between a box that does DNAT in the iptables `PREROUTING` chain, and a box that does DNAT in the nftables `prerouting` hook.

(Of course, Calico may also want to consider wholesale porting from iptables to nftables, but that's a separate thing...)

----

However, this particular section of the KEP is not talking about things like Calico, it's talking about whether we want to open up the black boxes (for the `nftables` backend), and let people poke around _inside_ them. We never did this for `iptables`, and as far as we know, no "real" products are trying to do this (other than trying to reuse `KUBE-MARK-MASQ`, which [we've now explicitly discouraged](https://kubernetes.io/blog/2022/09/07/iptables-chains-not-api/)). We have done a lot of refactoring of iptables rules in the last few years, and the corresponding PRs don't show any pingbacks from third-party projects saying things like "we need to update for the chain renaming in https://github.com/kubernetes/kubernetes/pull/109060", so I think we are actually OK there.

> 

Reviewer 8: This is now mentioned in risks and there is more detailed section of
"[Defining an API for integration with admin/debug/third-party rules" about it.

Resolving.

- We have at least the start of a plan for the next steps (changing
  the default mode, deprecating the old backends, etc).

### Upgrade / Downgrade Strategy

The new mode should not introduce any upgrade/downgrade problems,
excepting that you can't downgrade or feature-disable a cluster using
the new kube-proxy mode without switching it back to `iptables` or
`ipvs` first. (The older kube-proxy would refuse to start if given
`--proxy-mode nftables`, and wouldn't know how to clean up stale
nftables service rules if any were present.)

When rolling out or rolling back the feature, it should be safe to
enable the feature gate and change the configuration at the same time,
since nothing cares about the feature gate except for kube-proxy
itself. Likewise, it is expected to be safe to roll out the feature in
a live cluster, even though this will result in different proxy modes
running on different nodes, because Kubernetes service proxying is
defined in such a way that no node needs to be aware of the
implementation details of the service proxy implementation on any
other node.

### Version Skew Strategy

The feature is isolated to kube-proxy and does not introduce any API

> changes, so the versions of other components do not matter.

Author: There are no caveats; I mention it in the rollout section already (since you will necessarily run into this if you roll out the change in an existing cluster).

Yes, though it is necessary to clean up the nftables rules that were
created, or they will continue to intercept service traffic. In any
normal case, this should happen automatically when restarting
kube-proxy in `iptables` or `ipvs` mode, however:

  - If the user wants to roll back the cluster to a version of
    Kubernetes that doesn't have the nftables kube-proxy code (i.e.,
    rolling back from Alpha to Pre-Alpha), then they would need to
    make sure that the nftables rules got cleaned up _before_ they
    rolled back, since the older kube-proxy would not know how to
    clean them up.

  - If the user is rolling back in response to bugs in the `nftables`
    backend, then it is theoretically possible that the same bugs
    might affect the cleanup code, in which case it would be necessary
    to manually delete the nftables rules, or reboot or reprovision
    the node.

###### What happens if we reenable the feature if it was previously rolled back?

It should just work.

###### Are there any tests for feature enablement/disablement?

The actual feature gate enablement/disablement itself is not
interesting, since it only controls whether `--proxy-mode nftables`

> can be selected.

Author: Is that sort of approach to rollback requirements typical?

What if, say, someone upgrades to kube 1.31, decides to try using the now-beta nftables kube-proxy backend, then discovers a bug in _something else_ in 1.31 (storage backends, say) and decides to roll back to 1.30? Would we expect them to read the warnings about nftables rollback if nftables wasn't the reason they were rolling back?

> can be selected.

Author: makes sense

> can be selected.

Reviewer 8: Do you think it would make sense to summarize this comment in the KEP for the future (or maybe just link this comment thread)?

> can be selected.

Author: I was going to look through all the still-unresolved threads in this PR before beta. But sure, I can add it explicitly.

  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

TBD. It's not clear yet what sort of metrics will be interesting. For
example, in the `iptables` backend we have
`sync_proxy_rules_iptables_total`, which tells you the total number of
iptables rules kube-proxy has programmed. But the equivalent metric in
the `nftables` backend is not going to be as interesting, because many
of the things that are done with rules in the `iptables` backend will
be done with maps and sets in the `nftables` backend. Likewise, just
tallying "total number of rules and set/map elements" is not likely to
be useful, because the entire point of sets and maps is that they have
more-or-less **O(1)** behavior, so knowing the number of elements is
not going to give you much information about how well the system is

> likely to be performing.

Reviewer 8: Not blocking for alpha, but some performance metrics might be useful (NetworkProgramminLatency is something that I'm interested in)

Author: Yeah, NetworkProgrammingLatency is easy, and makes sense. Likewise SyncProxyRulesLastQueuedTimestamp/SyncProxyRulesLastTimestamp, and SyncProxyRulesLatency. (I actually want to move those out of the individual backends and into `pkg/proxy` as part of the cleanup/refactoring/librarification.)

It's more the replacements for the iptables-specific metrics that I'm unsure about. I'll clarify.

              "key": {
                "concat": [
                  {
                    "payload": {
                      "protocol": "ip",
                      "field": "daddr"
                    }
                  },
                  {
                    "payload": {
                      "protocol": "ip",
                      "field": "protocol"
                    }
                  },
                  {
                    "payload": {
                      "protocol": "th",
                      "field": "dport"
                    }
                  }
                ]
              },
              "data": "@service_ips"
            }
          }
        ]
      }
...
```

While it's clear how this _particular_ rule would be converted back
and forth between the two forms, there is no way to be able to map
_all_ rules back and forth without having separate code for every rule
type. Furthermore, the JSON syntax of individual rules is poorly
documented, and essentially all examples on the web (including the
nftables wiki, random blog posts, etc) use the non-JSON syntax. So if
we used the JSON syntax in kube-proxy, it would make the code harder

> to understand and to maintain.

Reviewer: what I'm still confused is if we can unequivocally can now if we can do a diff of `nft -f myrules.txt` is equal to `nft --json list table ip kube-proxy >myrules.json`, or we don't really need to compare the input against the output?

Author: You've got a bunch of extra or missing words there and I'm not sure exactly what you're asking, but anyway, there is no point where we need to actually look at the text of a rule that we previously created.

e.g., if we think that the `services` chain might contain out-of-date rules (because we just started up, and the existing chain might have been created by a previous version of kube-proxy), then there's no point in _looking at_ the chain to try to figure that out. We can just do:

```
	add chain ip kube-proxy services
	flush chain ip kube-proxy services
	add rule ip kube-proxy services ip daddr . ip protocol . th dport vmap @service-ips
	add rule ip kube-proxy services fib daddr type local fib oif != "lo" ip protocol . th dport vmap @service-nodeports
```

The first line ensures that the chain exists ("add" is a no-op if it already existed), the second line ensures that it is empty, and the third and fourth lines add the 2 rules that we want. This is executed atomically, so the net effect is to ensure that the chain exists and contains exactly the rules we want it to, regardless of what it contained before. And basically all updates work that way. (The iptables proxy works the same way.)

In some cases, we need to be able to get the names of all currently-existing chains and sets, so that in case there are stale ones, we can delete them. (Again, the iptables proxy does the same thing.) But that just requires parsing out the _names_ of chains, not their contents, and we can do that with the JSON API with no problem.

(For comparison, in https://github.com/containernetworking/plugins, there is code where we _do_ need to see if particular rules exist; in particular, the CNI plugins need to be able to distinguish between "there is a rule for `192.168.0.5` and it was created for container `foo`" and "there is a rule for `192.168.0.5` and it was created for container `bar`". The iptables code does this by adding a `--comment` to the rule, containing a hashed containerID. The nftables version will do the same thing, and again, we _can_ do that with the JSON API, because in nftables the comment on a rule is stored in a separate field from the expression, so we can list the rules with the JSON API and just look at the `"comment"` fields to identify them, and the `"handle"` fields to get a numeric ID to use to delete the rule if we don't want it.)

Reviewer: I apologize, I realize now  that I broke the sentence more than 4 times 😄 , let me start over. I was just wondering if we can diff the blob that we write with the one that we read and know the differences, instead of relying on the `nft` binary to do the magic 

> In some cases, we need to be able to get the names of all currently-existing chains and sets, so that in case there are stale ones, we can delete them. (Again, the iptables proxy does the same thing.) But that just requires parsing out the names of chains, not their contents, and we can do that with the JSON API with no problem.

 seems that based on this comment we do, indeed, I was worried about detecting stale chains

Author: > I was just wondering if we can diff the blob that we write with the one that we read and know the differences

In general, no, not really, because of syntactical ambiguities. eg, you might write a rule like:

```
    nft add rule ip kube-proxy postrouting  mark and 0x2000 == 0x2000  masquerade
```

but if you read that back you'll see

```
    chain postrouting {
        meta mark & 0x00002000 == 0x00002000 masquerade
    }
````

This is the other reason why not being able to translate rules back from the JSON form isn't a big deal: because even if you had them in the plaintext form, they might look unrecognizably different from what you were expecting.

> indeed, I was worried about detecting stale chains

Yeah, even in iptables, we don't do that by diffing the actual `iptables-save` and `iptables-restore` data against each other anyway; we parse out the chain names from the `iptables-save` output, and then compare that against another set we build up while generating the `iptables-restore` input.

`--nodeport-addresses` includes `127.0.0.1`), but _not_ change
`route_localnet`. In that case, we could document that administrators
could enable `route_localnet` themselves if they wanted to support
NodePorts on `127.0.0.1`, but then they would also be responsible for
mitigating any security holes they had introduced.

<<[/UNRESOLVED]>>
```

[CVE-2020-8558]: https://nvd.nist.gov/vuln/detail/CVE-2020-8558
[didn't also change certain other sysctls]: https://github.com/kubernetes/kubernetes/pull/91666#issuecomment-640733664
[may block legitimate traffic]: https://github.com/kubernetes/kubernetes/pull/91666#issuecomment-763549921
[contiv/vpp#1434]: https://github.com/contiv/vpp/pull/1434

#### NodePort Addresses

In addition to the localhost issue, iptables kube-proxy defaults to
accepting NodePort connections on all local IPs, which has effects
varying from intended-but-unexpected ("why can people connect to
NodePort services from the management network?") to clearly-just-wrong
("why can people connect to NodePort services on LoadBalancer IPs?")

The nftables proxy should default to only opening NodePorts on a
single interface, probably the interface with the default route by

> default. (Ideally, you really want it to accept NodePorts on the

Reviewer: Is here the problem the IPs or the interfaces? if one interface has multiple IPs we don't want to open it only on what we should consider one or all IPs?

Author: There seem to be two classes of complaint:
- I don't want NodePorts to work on `eth1`
- I don't want NodePorts to work on externalIPs / LB IPs

In the latter case, the IPs in question might be on the same interface as the primary node IP.

Reviewer 6: If we are revisiting this from scratch, let's dial it ALL the way back.  I like starting with the IP with default route.  If that is not sufficient, IPs matching a flag is pretty powerful.  If that is not sufficient, we can look at other options (same as detect-local-mode)

actually do need them, or if creating rules in another table is
sufficient.

```
<<[UNRESOLVED external rule integration API ]>>

It will be easier to figure out what the right thing to do here is
once we actually have a working implementation.

<<[/UNRESOLVED]>>
```

#### Rule monitoring

Proxies based on iptables have historically had the problem that
system processes (particularly firewall implementations) would
sometimes flush all iptables rules and restart with a clean state,
thus completely breaking kube-proxy. The initial solution for this
problem was to just recreate all iptables rules every 30 seconds even
if no services/endpoints had changed. Later this was changed to create
a single "canary" chain, and check every 30 seconds that the canary
had not been deleted, and only recreate everything from scratch if the
canary disappears.

NFTables provides a way to monitor for changes without doing polling;
you can keep a netlink socket open to the kernel (or a pipe open to an
`nft monitor` process) and receive notifications when particular kinds
of nftables objects are created or destroyed.

However, the "everyone uses their own table" design of nftables means

> that this may not be necessary. IPTables-based firewall

Author: There are a million things an administrator could do that would break kube-proxy; deleting the DaemonSet, removing the node IP, `down`ing the interface, adding bad routes, accidentally creating rules in other tables that act on k8s traffic, setting or unsetting various networking sysctls...

In the case of iptables, the issue was that if you used an iptables firewall _in the way that you were supposed to use it_, it would break kube-proxy. That was bad and we had to work around it. In the nftables case, there is no reason to think that this specific sort of breakage is going to happen, and thus no reason to pre-emptively try to protect against it.

> that this may not be necessary. IPTables-based firewall

Reviewer: Agree that you can not stop disasters like `sudo rm -f` , but disagree that `nftables` has enough guardrails to prevent problems , is not just `iptables firewalls`, are implementations of network policies or service meshes, or network plugins with portmap, ...

> that this may not be necessary. IPTables-based firewall

Reviewer: in order to keep our software reliable we should keep implementing the controller pattern with level based controllers https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md#writing-controllers
```
for {
  desired := getDesiredState()
  current := getCurrentState()
  makeChanges(desired, current)
}
```
and not depend on `nft` to do the right thing, the important is the sum of all parts, if we can detect and warn or repair a situation we should do it, if the is a bug in the other component this will allow us to signal and detect it without disruption

> that this may not be necessary. IPTables-based firewall

Author: > is not just iptables firewalls

No, it _is_ just iptables firewalls. There is a very specific problem there which we had to work around, and we work around it very specifically. (`utiliptables.Monitor` does not protect against arbitrary iptables changes by users; it _only_ attempts to catch the case where every iptables rule gets deleted at the same time.)

> implementations of network policies or service meshes, or network plugins implementing portmap

If any of those things touch our nftables rules, then they are unambiguously broken. And I don't even just mean "they're using the nftables API in a naughty and unsupported way", I mean "they can't possibly be getting the behavior they want", because if their functionality depends on them changing our rules, it will break when change those rules back (as we will _eventually_, even if we don't notice the changes immediately).

(To be clear: unless we add some API a la the ["admin override" maps](https://github.com/kubernetes/enhancements/pull/3824/files#diff-85ab4fa1a983d0d1d33b6bc089dc2c5c0b558d73672e7e2afad8a647ad210061R1168) that are vaguely suggested in the KEP, then we would not support any user/component _ever_ making _any_ changes to the `kube-proxy` table. You do not need to make changes or additions to the `kube-proxy` table to implement NetworkPolicy or portmaps; you make _your own_ tables and put your rules there.)

> in order to keep our software reliable we should keep implementing the controller pattern with level based controllers

Absolutely, we should not assume that kube-proxy manages to observe all events, or that it is always running.

But we intentionally don't stick _strictly_ to the level-based controller pattern in iptables kube-proxy, because, given the APIs available to us and the amount of data we have to process, that would result in unacceptable levels of performance. Instead, we do a slightly-looser sync loop, with periodic "overwrite everything with the expected state" fallbacks. The nftables backend will presumably work the same way.

(Or maybe the nftables APIs will turn out to be efficient enough that we can do a stricter sync instead... the big problem with iptables is that `getCurrentState()` is almost as slow as `makeChanges(desired, nil)` would be.)

> if we can detect and warn or repair a situation we should do it

There is a maintenance burden to having code to detect/warn/repair, so we shouldn't do it unless we think it is _actually_ going to be a problem. At the current time, we have no reason to believe that people deleting our nftables rules out from under us is going to be a problem.

> that this may not be necessary. IPTables-based firewall

Reviewer: > There is a maintenance burden to having code to detect/warn/repair, so we shouldn't do it unless we think it is actually going to be a problem. At the current time, we have no reason to believe that people deleting our nftables rules out from under us is going to be a problem.

I see, ok, agree with most of your points, I'm not advocating for doing the reconcile each time, but I really believe we should  have a repair loop that run every X mins that do the diff, like we do with the allocators, this loop has proven to be fantastic all these years to improve the resilience of the apiserver, if we add metrics to this loop like in https://github.com/kubernetes/kubernetes/pull/120843,  this will allow us to confirm with data the hypothesis that it is not really needed ...

> that this may not be necessary. IPTables-based firewall

Reviewer 3: Any repair / reconcile for **alpha** should aim to be as simple as we can make it (and any spare capacity dedicated to help with migration, docs, supporting tooling).

For beta and certainly before GA we should aim to make the reconcile mechanism work well even for high throughput and large scale - so no “fetch and compare all nftables records every inner invocation” shortcuts.

As this KEP PR aims to be `implementable` at alpha, I think we only need a nod / `UNRESOLVED` paragraph about what we might do for later phases. Let's [DoTheSimplestThingThatCouldPossiblyWork](https://wiki.c2.com/?DoTheSimplestThingThatCouldPossiblyWork) for our first pass at alpha.

> that this may not be necessary. IPTables-based firewall

Author: > I really believe we should have a repair loop that run every X mins that do the diff

We handle "repair"; we just don't do it with a separate loop or a diff. `syncProxyRules` just completely overwrites the current state with the desired state sometimes, to ensure that if any errors have crept in, they get fixed.

In the iptables code, there is a separate goroutine that polls the canary chains and forces an immediate "repair" (aka "full sync") if they disappear. (It has to do this in a slightly-weird way, to minimize the amount of time it holds the xtables lock, to avoid slowing down the main sync thread in the normal case where the canaries _haven't_ been deleted.) In the nftables code, I don't plan to have that check, since we don't expect that failure mode, but we'll still have the _periodic_ "completely overwrite the current state with the desired state", as well as the "completely overwrite the current state with the desired state" if we ever get back an error from nftables (eg it claims a chain doesn't exist when we think it should).

> that this may not be necessary. IPTables-based firewall

Author: OK, I think I see what you were asking in the other thread about reading back rules and diffing.

In the iptables backend, the only diff-like thing we ever do is that we read back the set of chain names that currently exist, so we can compare them against the set of chains we think _should_ exist, and delete any stale ones that were left behind.

Other than that, we _never_ compare the existing rules to the desired rules. If we want to make any change to an iptables chain, we just tell iptables "delete all of the existing rules in this chain and set the rules to be these new ones". (And if we _don't_ want to make a change to a chain, then we just ignore it, until either we get an error, or a timeout occurs, at which point we decide that we want to change _every_ chain all at once.)

(The only reason we don't do the same thing for deleting stale chains is that iptables doesn't give us any API that would let us do that, unless we wanted to delete all non-Kubernetes chains too.)

The nftables backend will likely work exactly the same, in this regard, except that there may be more things we need to delete besides just chains (eg, sets and maps).

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

- <test>: <link to test coverage>

#### Scalability & Performance tests

```
<<[UNRESOLVED perfscale ]>>

- For the control plane side, the existing scalability tests are
  probably reasonable, assuming we implement the same
  `NetworkProgrammingLatency` metric as the existing backends.

- For the data plane side, there are tests of
  `InClusterNetworkLatency`, but no one is really looking at the
  results yet and they may need work before they are useable.

- We should also make sure that other metrics (CPU, RAM, I/O, etc)
  remain reasonable in an `nftables` cluster.

<<[/UNRESOLVED]>>
```

### Graduation Criteria

#### Alpha

- `kube-proxy --proxy-mode nftables` available behind a feature gate

> - nftables mode has unit test parity with iptables

Reviewer 8: What is the test coverage of iptables?

Playing devils advocate a bit, if it would be 10%, then it's not good enough criteria.
Can you please fill in the test coverage for iptables/ipvs (there are links in KEP template) and then potentially adjust here if needed?

Author: (https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit does not seem to be loading for me? I get the header but then it just says "Loading results from server..." forever. Or does it just take a really really really long time?)

But from KEP-3453:

```
- `k8s.io/kubernetes/pkg/proxy/iptables`: `2022-08-02` - `67%`
```

and it's probably still within 10% of that

Reviewer: from current master
```
coverage: 70.6% of statements
ok  	k8s.io/kubernetes/pkg/proxy/iptables	0.364s	coverage: 70.6% of statements
```

Reviewer 8: @danwinship - can you please paste the above to the KEP?

Now, getting back to the core of my question - I understand that we generally won't be touching iptables backend so requiring changes there would be counter-productive.

But given it will be replacement for the existing things, and as identifier by @liggitt those are things where we observe the biggest number of regressions, I think that our bar for coverage for nftables should actually be higher.

WDYT?

Author: I already did add the iptables number to [the "Unit tests" section](https://github.com/kubernetes/enhancements/pull/3824/files#diff-85ab4fa1a983d0d1d33b6bc089dc2c5c0b558d73672e7e2afad8a647ad210061R1321)

The rest of your comment seems definitely post-alpha?

Reviewer 8: > The rest of your comment seems definitely post-alpha?

For unit tests I would actually expect those coming together with the implementation.

For all other tests, having them post alpha makes sense.

- An nftables-mode e2e job exists, and passes

- Documentation describes any changes in behavior between the
  `iptables` and `ipvs` modes and the `nftables` mode.

- Documentation explains how to manually clean up nftables rules in
  case things go very wrong.

#### Beta

- At least two releases since Alpha.

> - The nftables mode has seen at least a bit of real-world usage.

Author: Not specifically, no. (I adding some fudging here with "at least a bit".)

I guess if we fail to get any real-world usage and no one dares to try it until we declare it beta, then we can amend the KEP at that point.

> - The nftables mode has seen at least a bit of real-world usage.

Reviewer: from current master
```
coverage: 70.6% of statements
ok  	k8s.io/kubernetes/pkg/proxy/iptables	0.364s	coverage: 70.6% of statements
```

- No major outstanding bugs.

- A "kube-proxy mode-switching" e2e job exists, to confirm that you
  can redeploy kube-proxy in a different mode in an existing cluster.
  Rollback is confirmed to be reliable.

- An nftables e2e periodic perf/scale job exists, and shows
  performance as good as iptables and ipvs.

- Documentation describes any changes in behavior between the

>   `iptables` and `ipvs` modes and the `nftables` mode. Any warnings

Reviewer: any examples to the kind of metrics you are looking for?

>   `iptables` and `ipvs` modes and the `nftables` mode. Any warnings

Author: @aojea eg, we could add up the packet counters on all of the localhost nodeport rules whenever we do `iptables-save`, and then expose that number as a metric. And then we could tell people "if your localhost_nodeport_connections metric is non-0, then something in your cluster depends on localhost nodeports working, so you can't migrate to nftables without fixing that first". etc

>   `iptables` and `ipvs` modes and the `nftables` mode. Any warnings

Reviewer 8: Yes - this is a perfect example of thing that we should have (for Beta).

  that we have decide to add for `iptables` users using functionality
  that behaves differently in `nftables` have been added.

- No UNRESOLVED sections in the KEP. (In particular, we have figured
  out what sort of "API" we will offer for integrating third-party
  nftables rules.)

#### GA

> - At least two releases since Beta.

Reviewer: we have to ensure all the nftables features used are present in at least the main distros, no? there is an open issue related about minimum kernel version required https://github.com/kubernetes/kubernetes/issues/116799

Reviewer: commented here too https://github.com/kubernetes/enhancements/pull/3824/files#r1326450345, closing as dup

changes anyway.)

### We will hopefully be able to trade 2 supported backends for 1

Right now SIG Network is supporting both the `iptables` and `ipvs`
backends of kube-proxy, and does not feel like it can ditch `ipvs`
because of perceived performance issues with `iptables`. If we create a new
backend which is as functional and non-buggy as `iptables` but as
performant as `ipvs`, then we could (eventually) deprecate both of the
existing backends and only have one Linux backend to support in the future.

### Writing a new kube-proxy mode will help to focus our cleanup/refactoring efforts

There is a desire to provide a "kube-proxy library" that third parties
could use as a base for external service proxy implementations
([KEP-3786]). The existing "core kube-proxy" code, while functional,
is not very well designed and is not something we would want to
support other people using in its current form.

Writing a new proxy backend will force us to look over all of this
shared code again, and perhaps give us new ideas on how it can be
cleaned up, rationalized, and optimized.

[KEP-3786]: https://github.com/kubernetes/enhancements/issues/3786

> ### Goals

Reviewer 6: Can we add a goal like "Design and document where and how external actors can plug into the packet processing logic safely".  We never made this easy in iptables mode, e.g. if I wanted to `-j LOG` for a particular service.

Reviewer 6: I see this below, but maybe a nod here?

Author: It's there in Goals:

```
    - Document specific details of the nftables implementation that we
      want to consider as "API". In particular, document the
      high-level behavior that authors of network plugins can rely
      on. We may also document ways that third parties or
      administrators can integrate with kube-proxy's rules at a lower
      level.
```

### Risks and Mitigations

#### Functionality

The primary risk of the proposal is feature or stability regressions,
which will be addressed by testing, and by a slow, optional, rollout
of the new proxy mode.

The most important mitigation for this risk is ensuring that rollback
from `nftables` mode back to `iptables`/`ipvs` mode works reliably.

#### Compatibility

Many Kubernetes networking implementations use kube-proxy as their
service proxy implementation. Given that few low-level details of
kube-proxy's behavior are explicitly specified, using it as part of a
larger networking implementation (and in particular, writing a
NetworkPolicy implementation that interoperates with it correctly)
necessarily requires making assumptions about (currently-)undocumented
aspects of its behavior (such as exactly when and how packets get
rewritten).

While the `nftables` mode is likely to look very similar to the
`iptables` mode from the outside, some CNI plugins, NetworkPolicy
implementations, etc, may need updates in order to work with it. (This
may further limit the amount of testing the new mode can get during
the Alpha phase, if it is not yet compatible with popular network

> plugins at that point.)

Reviewer 8: Thanks for adding this - do we have some mitigation here that we want to mention?

Author: not really; we can try to avoid adding making it *gratuitously* different, but we can't avoid the fact that it's going to be different.

(This is a separate "don't be too different" concern from the user-level behavioral differences.)

based on destination IP, protocol, and port. We will then need only a
single actual rule to apply the verdict map to all inbound traffic.
(Or it may end up making more sense to have separate verdict maps for
ClusterIP, ExternalIP, and LoadBalancer IP?) Either way, service
dispatch will be roughly **O(1)** rather than **O(n)** as in the
`iptables` backend.

Likewise, for NodePort traffic, we will use a verdict map matching
only on destination protocol / port, with the rules set up to only
check the `nodeports` map for packets addressed to a local IP.

```
map service_ips {
  comment "ClusterIP, ExternalIP and LoadBalancer IP traffic";

  # The "type" clause defines the map's datatype; the key type is to
  # the left of the ":" and the value type to the right. The map key
  # in this case is a concatenation (".") of three values; an IPv4
  # address, a protocol (tcp/udp/sctp), and a port (aka
  # "inet_service"). The map value is a "verdict", which is one of a
  # limited set of nftables actions. In this case, the verdicts are
  # all "goto" statements.

  type ipv4_addr . inet_proto . inet_service : verdict;

  elements {

>     172.30.0.44 . tcp . 80 : goto svc_4SW47YFZTEDKD3PK,

Reviewer 6: are names as restricted as in iptables (30 chars)?  Or can we make names more human friendly?

Author: I was going to say "yes", but based on some testing, the limit is something higher than 128 chars...

set need_masquerade {
  type ipv4_addr . inet_service . ipv4_addr . inet_service . inet_proto;
  flags timeout ; timeout 5s ;
}

chain mark_for_masquerade {
  update @need_masquerade { ip saddr . th sport . ip daddr . th dport . ip protocol }
}

chain postrouting_do_masquerade {
  # We use "ct original ip daddr" and "ct original proto-dst" here
  # since the packet may have been DNATted by this point.

  ip saddr . th sport . ct original ip daddr . ct original proto-dst . ip protocol @need_masquerade masquerade
}
```

This is not yet tested, but some kernel nftables developers have
confirmed that it ought to work. We should test to make sure that
having a potentially-high-churn `need_masquerade` set will not be a
performance problem.

##### Session affinity

Session affinity can be done in roughly the same way as in the
`iptables` proxy, just using the more general nftables "set" framework
rather than the affinity-specific version of sets provided by the
iptables `recent` module. In fact, since nftables allows arbitrary set
keys, we can optimize relative to `iptables`, and only have a single
affinity set per service, rather than one per endpoint. (And we also

> have the flexibility to change the affinity key in the future if we

Reviewer 6: Yeah, we should queue up the discussion about having a different affinity key - iptables way vs ipvs way.

since localhost NodePorts do not work with `ipvs` mode or with IPv6,
many places that used to use NodePorts on `127.0.0.1` have already
been rewritten to not do so (eg [contiv/vpp#1434]).

So:

  - There is no way to make IPv6 localhost NodePorts work with a
    NAT-based solution.

  - The way to make IPv4 localhost NodePorts work with NAT introduces
    a security hole, and we don't necessarily have a fully-generic way
    to mitigate it.

  - The only commonly-argued-for use case for the feature involves
    deploying a service in a configuration which its own documentation
    describes as insecure and "only appropriate for testing".

      - The use case in question works by default against cri-dockerd
        but not against containerd or cri-o with their default
        configurations.

      - cri-dockerd, containerd, and cri-o all allow additional
        "insecure registry" IPs/CIDRs to be configured, so an
        administrator could configure them to allow non-TLS image
        pulling against a ClusterIP.

Given this, I think we should not try to support localhost NodePorts
in the `nftables` backend.

```
<<[UNRESOLVED dnat-but-no-route_localnet ]>>

> As a possible compromise, we could make the `nftables` backend create

Reviewer 6: let's keep this in our pocket, in case a real use-case emerges, but not implement until then?

<<[/UNRESOLVED]>>
```

[CVE-2020-8558]: https://nvd.nist.gov/vuln/detail/CVE-2020-8558
[didn't also change certain other sysctls]: https://github.com/kubernetes/kubernetes/pull/91666#issuecomment-640733664
[may block legitimate traffic]: https://github.com/kubernetes/kubernetes/pull/91666#issuecomment-763549921
[contiv/vpp#1434]: https://github.com/contiv/vpp/pull/1434

#### NodePort Addresses

In addition to the localhost issue, iptables kube-proxy defaults to
accepting NodePort connections on all local IPs, which has effects
varying from intended-but-unexpected ("why can people connect to
NodePort services from the management network?") to clearly-just-wrong
("why can people connect to NodePort services on LoadBalancer IPs?")

The nftables proxy should default to only opening NodePorts on a
single interface, probably the interface with the default route by
default. (Ideally, you really want it to accept NodePorts on the
interface that holds the route to the cloud load balancers, but we
don't necessarily know what that is ahead of time.) Admins can use
`--nodeport-addresses` to override this.

#### Behavior of service IPs

```
<<[UNRESOLVED unused service IP ports ]>>

> @thockin has suggested that service IPs should reject connections on

Reviewer 6: We should reject or drop, but almost certainly not ignore such connections.  I lean towards reject, but I can see arguments in both directions.

Reviewer: reject improves supportability, the error is more meaningful "connection refused" indicates someone does not want you to go through , drop is "connection timeout" 
 I've seen thousands of those red herrings in customers tickets at SUSE, at RedHat and at Google now, it is easy to write a playbook as "connect to a service and connection refused check if there are endpoints", however, "connect timeout check if there are endpoints and if everything in the path is not filtering the packet"

ports they aren't using. (This would most easily be implemented by

> adding a `--service-cidr` flag to kube-proxy so we could just "reject

Reviewer 6: the less things that we teach about things like service-cidr the better.

Author: with KEP-1880 kube-proxy could just watch the ServiceCIDR objects, and no one would need to explicitly tell it anything

everything else", but even without that we could at least reject
connections on inactive ports of active service IPs.)

<<[/UNRESOLVED]>>
```

```
<<[UNRESOLVED service IP pings ]>>

Users sometimes get confused by the fact that service IPs do not
respond to ICMP pings, and perhaps this is something we could change.

<<[/UNRESOLVED]>>
```

> #### Defining an API for integration with admin/debug/third-party rules

Reviewer 6: +1 to this!
